### PR TITLE
Add support for System.DateOnly (#1303)

### PR DESCRIPTION
**Pull request by SiCoe**
This is a draft pull request relating to the descussion: [Support `DateOnly` and `TimeOnly` by default for .NET 6+](https://github.com/AutoFixture/AutoFixture/discussions/1304).

Currently only `DateOnly` is implemented because I want to wait on feedback on adding net6.0 as a TargetFramework.

### CODE REVIEW

**Comment by Kralizek** [comment]
Wouldn't it be better to just use the symbol `NET6_0_OR_GREATER` instead of defining a new one for every feature we want to support?

**Comment by SiCoe** [comment]
> 
> 
> Wouldn't it be better to just use the symbol `NET6_0_OR_GREATER` instead of defining a new one for every feature we want to support?

I started with using the built in `NET6_0` and swapped it out for custom symbol as didn't want to lock it to specifically .NET 6.0 for if new targets get added in the future. I think I do prefer `NET6_0_OR_GREATER` though, as can be used if other features require it.

I'll swap it over.

**Comment by Kralizek** [comment]
I agree with you, `NET6_0` would have been very binding. But unless Microsoft starts forking .NET in multiple runtimes with different capabilities again (which I highly doubt), `NET6_0_OR_GREATER` should work pretty well.

**Comment by aivascu** [comment]
@SiCoe I am having issues running the build on the entire solution with the net6 sdk. Is the build running for you?

File: Src/Common.props [hunk 1]
```diff
@@ -54,7 +54,7 @@
     <DefineConstants Condition=" '$(TargetFramework)'=='netstandard2.0' Or '$(TargetFramework)'=='netcoreapp2.1' Or '$(TargetFramework)'=='net6.0' ">$(DefineConstants);SYSTEM_NET_MAIL</DefineConstants>
   </PropertyGroup>
   <PropertyGroup>
-    <DefineConstants Condition=" '$(TargetFramework)'=='net6.0' ">$(DefineConstants);SYSTEM_DATEONLY</DefineConstants>
+    <DefineConstants Condition=" '$(TargetFramework)'=='net6.0' ">$(DefineConstants);NET6_0_OR_GREATER</DefineConstants>
```
Comment (Kralizek) [inline Src/Common.props:57]: No need to define this yourself. the SDK will take care of defining these symbols

https://docs.microsoft.com/en-us/dotnet/standard/frameworks#how-to-specify-a-target-framework

Comment (SiCoe) [reply Src/Common.props]: Thank you, I was unaware the SDK did the `_OR_GREATER` constants. Changed.

Comment (Kralizek) [reply Src/Common.props]: We just need to make sure that we use a compatible .NET SDK in the CI pipeline

@aivascu

**Comment by SiCoe** [comment]
> 
> 
> @SiCoe I am having issues running the build on the entire solution with the net6 sdk. Is the build running for you?

I've been using the `Src/AutoFixture.slnf` which has been building fine. If I open `Src/All.sln` I do get compliation errors, something I'll need to look into.

I think AppVeyor will need updating too

**Comment by aivascu** [comment]
Looks like FSharp.Core doesn't resolve properly for F# projects, under .NET 6 SDK, even though the versions are locked. Seems to me, it is a bug in the SDK. I get one of these each time Microsoft decides to update the SDK.
Raising the minimum version to 4.2.3 and greater seems to solve the issue. I would prefer not to raise the minimum version for F#, unless the code requires some special features, but it seems I won't have a choice if I don't find another fix.

**Comment by SiCoe** [comment]
> Looks like FSharp.Core doesn't resolve properly for F# projects, under .NET 6 SDK, even though the versions are locked. Seems to me, it is a bug in the SDK. I get one of these each time Microsoft decides to update the SDK. Raising the minimum version to 4.2.3 and greater seems to solve the issue.

As you suggested, upgrading the FSharp.Core minimum package version to 4.2.3 does allow the solution to build successfully. (I've created a [new branch](https://github.com/SiCoe/AutoFixture/tree/feature/date-only-f%23-upgrade) with `DateOnly` implemented and the [FSharp.Core upgrade](https://github.com/SiCoe/AutoFixture/commit/dcb20392691816b50967854d9c1ed2e0e4291167).

> I would prefer not to raise the minimum version for F#, unless the code requires some special features, but it seems I won't have a choice if I don't find another fix.

I put the upgrade commit on a different branch for the moment to allow for you to try and find another fix. Maybe building different projects using different versions of the SDK could be an option? (Apologies for not offering more help; the SDK upgrade implications F# is not something I'm familiar with).

I'm happy to spend the time also implementing `TimeOnly` for this PR if you don't think the SDK issue will block this.

**Comment by dutts** [comment]
Is there a pre-release version somewhere that currently supports DateOnly? It's literally all that's stopping me from being able to use AutoFixture in my project to create some test instances of a rather large and deeply nested object.

**Comment by Kralizek** [comment]
I guess it will be released in the next major version. I'm not a maintainer so I can't anticipate any timeline.

I suggest you copy the bits you need and add them manually to your test project and remove them when AF5 is out.

**Comment by adamrodger** [comment]
In case anyone lands here from Google and this PR isn't merged yet, you can do this to get `DateOnly` support:

```csharp
fixture.Customize<DateOnly>(composer => composer.FromFactory<DateTime>(DateOnly.FromDateTime));
```

**Comment by hughesjs** [comment]
> ```cs
> fixture.Customize<DateOnly>(composer => composer.FromFactory<DateTime>(DateOnly.FromDateTime));
> ```

Legend...

**Comment by JeanetteNielsen** [comment]
@aivascu 
v6.0 seems a bit far in the future. Do you have a timeline? Is there anything keeping this from being merged in a sooner release? (I'm not that familiar with with your milestones and version challenges)

Unfortunetly, the suggestions to add it manually for test projects is very costly for well tested microservice setups. For us this will likely mean we need to consider moving away from AutoFixture.

**Comment by aivascu** [comment]
@JeanetteNielsen thank you for contacting me.

There is currently no timeline for v6, and due to technical limitations of the CI/CD pipeline I'll not be able to integrate .NET 6 before v5 is delivered. You can track the progress for v5 in #1327.

At this point I can only suggest to use the workaround specified in this thread and/or wait for v6.

I have resumed the work on v5 this month, after a 6 months break due to personal issues, and I plan to release the RC version in a couple months. This estimate however can change at any time depending on my availability. Please keep in mind that this project is not sponsored by anyone except the community's contributions and my will to spend my spare time working on it.

Also if you'll inspect the [license](https://github.com/AutoFixture/AutoFixture/blob/master/LICENCE.txt) under which AutoFixture is released, you'll see there is no warranty for fitness for any particular purpose, including microservices. In case you'll decide to move away it will be regrettable, and when/if you'll decide to use AutoFixture again, it will still be here available to anyone free of charge.

**Comment by JeanetteNielsen** [comment]
> @JeanetteNielsen thank you for contacting me.
> 
> There is currently no timeline for v6, and due to technical limitations of the CI/CD pipeline I'll not be able to integrate .NET 6 before v5 is delivered. You can track the progress for v5 in #1327.
> 
> At this point I can only suggest to use the workaround specified in this thread and/or wait for v6.
> 
> I have resumed the work on v5 this month, after a 6 months break due to personal issues, and I plan to release the RC version in a couple months. Work on v6 will begin immediately after v5 is released. These estimates however can change at any time depending on my availability or other circumstances. Please keep in mind that this project is not sponsored by anyone except the community's contributions and my will to spend my spare time working on it.
> 
> Also if you'll inspect the [license](https://github.com/AutoFixture/AutoFixture/blob/master/LICENCE.txt) under which AutoFixture is released, you'll see there is no warranty for fitness for any particular purpose, including microservices. In case you'll decide to move away it will be regrettable, and when/if you'll decide to use AutoFixture again, it will still be here available to anyone free of charge.

I'm definitely not expecting any warrenty, I simply mention microservice since it's a popular pattern, and is likely that quite a few others to have the same challenge.
Having that in mind, might have been useful in prioritizing feature requests, if the .net upgrade was not the issue.

Taking the time to comment is a sign of engagement and appreciation for your work. It might seem like nagging, but it's based on pure interest and a hope on continuing using the tool. 😊
Im not hanging around in your repo daily, so I'm not familiar with the amount of contributers or challenges you face.

**Comment by mrukas** [comment]
Thanks @adamrodger for the workaround. I really like Autofixture and it would be great to see this merged into the project because DateOnly is around for a while now.
Thanks to all the contributors.

**Comment by smolsky** [comment]
> In case anyone lands here from Google and this PR isn't merged yet, you can do this to get `DateOnly` support:
> 
> ```cs
> fixture.Customize<DateOnly>(composer => composer.FromFactory<DateTime>(DateOnly.FromDateTime));
> ```

@adamrodger, Thanks for solution!
Anyone who wants to see this workarround in action please follow the link: https://stackoverflow.com/questions/73732174/does-autofixture-support-dateonly-for-net6/

**Comment by swettstein** [comment]
> In case anyone lands here from Google and this PR isn't merged yet, you can do this to get `DateOnly` support:
> 
> ```cs
> fixture.Customize<DateOnly>(composer => composer.FromFactory<DateTime>(DateOnly.FromDateTime));
> ```

i had this workaround in place but noticed my tests were failing much more frequently because i was expecting unique dates in a collection. the RandomGenerator for DateTime is likely to generate a different DateTime for a collection of DateTime but when you pull back to DateOnly, it's much more likely to cause the same DateOnly to be generated. i ended up copying the DateTimeRandomGenerator and converting it to DateOnlyRandomGenerator (just like this solution in this PR) and it's much more likely to produce unique values in a collection of DateOnly.

i also created an Options class to manipulate the generated dates a little more. only caveat is that if you specify NoPastDates = true and NoFutureDates = true, you will end up with the TargetDate always.

```csharp
using System;
using System.Reflection;
using AutoFixture;
using AutoFixture.Kernel;

public class DateOnlyGeneratorOptions
{
    public int TimeSpanInYears { get; set; } = 4;
    public bool NoFutureDates { get; set; }
    public bool NoPastDates { get; set; }
    public DateOnly TargetDate { get; set; } = DateOnly.FromDateTime(DateTime.Today);
}

public class RandomDateOnlySequenceGenerator : ISpecimenBuilder
{
    private readonly RandomNumericSequenceGenerator _randomizer;

    public RandomDateOnlySequenceGenerator() : this(new DateOnlyGeneratorOptions())
    {

    }

    public RandomDateOnlySequenceGenerator(DateOnlyGeneratorOptions options) : this(
        options.TargetDate.AddYears(options.NoPastDates ? 0 : options.NoFutureDates ? -options.TimeSpanInYears : -(options.TimeSpanInYears / 2)),
        options.TargetDate.AddYears(options.NoFutureDates ? 0 : options.NoPastDates ? options.TimeSpanInYears : options.TimeSpanInYears / 2))
    {

    }

    public RandomDateOnlySequenceGenerator(DateOnly minDate, DateOnly maxDate)
    {
        if (minDate >= maxDate)
            throw new ArgumentException("The 'minDate' argument must be less than the 'maxDate'.");
        _randomizer = new RandomNumericSequenceGenerator(new long[]
        {
            minDate.DayNumber,
            maxDate.DayNumber
        });
    }

    public object Create(object request, ISpecimenContext context)
    {
        if (context == null)
            throw new ArgumentNullException(nameof(context));
        return IsDateOnlyRequest(request) ? CreateRandomDate(context) : new NoSpecimen();
    }

    private bool IsDateOnlyRequest(object request) => typeof(DateOnly).GetTypeInfo().IsAssignableFrom(request as Type);

    private object CreateRandomDate(ISpecimenContext context) => DateOnly.FromDayNumber(GetRandomDayNumber(context));

    private int GetRandomDayNumber(ISpecimenContext context) => (int)_randomizer.Create(typeof(int), context);
}
```

**Comment by aivascu** [comment]
The PR will not be merged in the current release version and neither in the v5 as the current CI pipeline does not support .NET 6. The feature will be moved to v6 in which we'll add support for .NET 6.
I'll close the PR for now and will reopen it once the project infrastructure is ready for it.

**Comment by mxa0079** [comment]
This should not be closed as it gives the impression that it was fixed. I suggest to keep it open and maybe close it for comments?

**Comment by guipcarvalho** [comment]
Where are we, with this PR? This issue is very annoying and a few years have passed.

**Comment by aivascu** [comment]
@guipcarvalho the reason this PR or any other PR on `DateOnly` will not be merged into v4 master is described in my comment above https://github.com/AutoFixture/AutoFixture/pull/1305#issuecomment-1657997482. The current v5 preview version already contains support for `DateOnly`. If you have any issues with it feel free to open a separate issue.

**Pull request closed without merging**